In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
df = pd.read_csv(r"C:\Users\ashis\OneDrive\Desktop\Indian Startup funding\uncleaned_startup_funding.csv")

In [5]:
df.head()

Sr No Date dd/mm/yyyy                  Startup Name    Industry Vertical  \
0      1      09-01-2020                        BYJU’S               E-Tech   
1      2      13-01-2020                        Shuttl       Transportation   
2      3      09-01-2020                     Mamaearth           E-commerce   
3      4      02-01-2020  https://www.wealthbucket.in/              FinTech   
4      5      02-01-2020                        Fashor  Fashion and Apparel   

                             SubVertical City  Location  \
0                             E-learning      Bengaluru   
1              App based shuttle service        Gurgaon   
2  Retailer of baby and toddler products      Bengaluru   
3                      Online Investment      New Delhi   
4            Embroiled Clothes For Women         Mumbai   

              Investors Name       InvestmentnType Amount in USD Remarks  
0    Tiger Global Management  Private Equity Round  20,00,00,000     NaN  
1  Susquehanna Growth Equity              Series C     80,48,394     NaN  
2      Sequoia Capital India              Series B   1,83,58,860     NaN  
3             Vinod Khatumal          Pre-series A     30,00,000     NaN  
4    Sprout Venture Partners            Seed Round     18,00,000     NaN

In [6]:
df.shape

(3044, 10)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3044 entries, 0 to 3043
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Sr No              3044 non-null   int64 
 1   Date dd/mm/yyyy    3044 non-null   object
 2   Startup Name       3044 non-null   object
 3   Industry Vertical  2108 non-null   object
 4   SubVertical        2873 non-null   object
 5   City  Location     3035 non-null   object
 6   Investors Name     3020 non-null   object
 7   InvestmentnType    3040 non-null   object
 8   Amount in USD      2084 non-null   object
 9   Remarks            419 non-null    object
dtypes: int64(1), object(9)
memory usage: 237.9+ KB


<h3>Renaming the columns</h3>

In [8]:
df.rename(columns={'Sr No':'sr_no', 
                   'Date dd/mm/yyyy':'date',
                  'Startup Name':'startup',
                  'Industry Vertical':'industry',
                  'SubVertical':'subvertical',
                  'City  Location':'city',
                  'Investors Name':'investors',
                  'InvestmentnType':'investment_type',
                  'Amount in USD':'amount',
                  'Remarks':'remarks'},inplace=True)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3044 entries, 0 to 3043
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   sr_no            3044 non-null   int64 
 1   date             3044 non-null   object
 2   startup          3044 non-null   object
 3   industry         2108 non-null   object
 4   subvertical      2873 non-null   object
 5   city             3035 non-null   object
 6   investors        3020 non-null   object
 7   investment_type  3040 non-null   object
 8   amount           2084 non-null   object
 9   remarks          419 non-null    object
dtypes: int64(1), object(9)
memory usage: 237.9+ KB


In [10]:
# Number of Null values in each column
df.isnull().sum()

sr_no                 0
date                  0
startup               0
industry            936
subvertical         171
city                  9
investors            24
investment_type       4
amount              960
remarks            2625
dtype: int64

1. date:-
      1. Datatype is of Object Type
      2. Inconsistent format in dates 
      3. Few encoding issues \\xc2\\xa010/7/2015 at sr_no 2607
      4. No null value

2. amount:-
      1. Data type Object
      2. Comma separated values
      3. 960 null values
      4. Encoding issues -- \\xc2\\xa0685,000
      5. undisclosed, unknown, N/A as entries

3. remarks:-
      1. Around 2600 Null values better to drop the column.

4. startup:-
      1. Many names has .com or .in or urls as name https://www.wealthbucket.in/
      2. Encoding issues \\xc2\\xa0News in shorts,
                         NearBuy (previously\\xc2\\xa0 groupon India),
                         Carl\\xe2\\x80\\x99s Jr,
                         Byju\\xe2\\x80\\x99s
                         \\xc2\\xa0CloudCherry
                         Let\xe2\x80\x99s Barter
                         "BYJU\\'S"
                         Alef\\nMobitech
      4. "/" in Names
                   -- Transtutor/Askiitians --> new name/old name
                   -- Zeolr/RespirON
                   -- The/Nudge Foundation --> This an actual name of the startup-- needs to be handled differently
         
      5. Many starup names have entries with different casings -- can be solved by converting in lowercase
      6. Different spellings (m.paani, mPaani), (crown-it, crownit), (oyo rooms, oyorooms)

5. city:-
      1. Encoding issues -- \\\\xc2\\\\xa0Noida
      2. Different names of same cities - gurgaon and gurugram, Bengaluru and Bangalore
      3. Area names in same city like Kormangla in Bengaluru
      4. 4 nan values
6. investment_type:-
      1. 3 Nan values
      2. encoding issues Seed\\\\nFunding
      3. Multiple variations of same type of investment
         
7. investors:-
      1. 20 nan values
      2. encoding
      3. multiple investors in one row
         
8. industry and subvertical:-
      1. around 600 nan values
      2. encoding issues

<h3>sr_no column</h3>

In [11]:
# dropping the sr_no column
df.drop(columns=['sr_no'], inplace=True)

<h3>remarks column</h3>

In [12]:
# dropping the remarks column
df.drop(columns=['remarks'],inplace=True)

In [13]:
print(df.columns)

Index(['date', 'startup', 'industry', 'subvertical', 'city', 'investors',
       'investment_type', 'amount'],
      dtype='object')


<h3> date column</h3>

In [14]:
# Taking care of the inconsistent format
df['date'] = df['date'].str.replace(r'\.','-')
df['date'] = df['date'].str.replace(r'\\','-')

In [15]:
# making sure if changes has been sucessfully made
df['date'][df['date'].str.contains(r'\\')]

Series([], Name: date, dtype: object)

In [16]:
# converting the date column to datetime
df['date']=pd.to_datetime(df['date'],dayfirst=True,errors='coerce')

In [17]:
df['date'].isnull().sum()

8

In [18]:
# filling up the nan values using bfill()
df['date']=df['date'].bfill()

In [19]:
df['date'].isnull().sum()

0

<h3>amount column</h3>

In [20]:
# converting entries with undisclosed, unknown and N/A into np.nan
# using replace you can only replace a string with a string not some other data type value
df['amount'] = df['amount'].replace(['Unknown','Undisclosed','N/A'],np.nan)

In [21]:
# removing the commas
df['amount'] = df['amount'].str.replace(',','')

In [22]:
# Dropping all rows with nan values in amount column
df.dropna(subset=['amount'],inplace=True)

In [23]:
# All the non digit entries in amount column
df['amount'][~df['amount'].str.isdigit()]

20             undisclosed
34                 unknown
76              4889975.54
81              2739034.68
83             15109500.00
89             undisclosed
91             undisclosed
109              14342000+
2602    \\xc2\\xa020000000
2603    \\xc2\\xa016200000
2604         \\xc2\\xa0N/A
2605         \\xc2\\xa0N/A
2606      \\xc2\\xa0600000
2607      \\xc2\\xa0685000
2608    \\xc2\\xa019350000
2609     \\xc2\\xa05000000
2610    \\xc2\\xa010000000
2611         \\xc2\\xa0N/A
2613         \\xc2\\xa0N/A
Name: amount, dtype: object

In [24]:
# removing the wrong encodings --> \\xc2\\xa0, + sign
df['amount'] = df['amount'].str.replace(r'\\xc2\\xa0','')
df['amount'] = df['amount'].str.replace('+','')

In [25]:
df['amount'][~df['amount'].str.isdigit()]

20      undisclosed
34          unknown
76       4889975.54
81       2739034.68
83      15109500.00
89      undisclosed
91      undisclosed
2604            N/A
2605            N/A
2611            N/A
2613            N/A
Name: amount, dtype: object

In [26]:
df['amount'] = df['amount'].replace(['unknown','undisclosed','N/A'],np.nan)

In [27]:
df['amount'].isnull().sum()

8

In [28]:
# Dropping all rows with nan values in amount column
df.dropna(subset=['amount'],inplace=True)

In [29]:
df['amount'].isnull().sum()

0

In [30]:
df['amount'] = df['amount'].astype(float).round()

In [31]:
# converting the data type of amount column to int
df['amount'] = df['amount'].astype(np.int32)

In [32]:
df['amount'].dtype

dtype('int32')

<h3>city column</h3>

In [33]:
df['city'].unique()

array(['Bengaluru', 'Gurgaon', 'New Delhi', 'Mumbai', 'Chennai', 'Pune',
       'Noida', 'Faridabad', 'San Francisco', 'San Jose,', 'Amritsar',
       'Delhi', 'Kormangala', 'Tulangan', 'Hyderabad', 'Burnsville',
       'Menlo Park', 'Gurugram', 'Palo Alto', 'Santa Monica', 'Singapore',
       'Taramani', 'Andheri', 'Chembur', 'Nairobi', 'Haryana', 'New York',
       'Karnataka', 'Mumbai/Bengaluru', 'Bhopal',
       'Bengaluru and Gurugram', 'India/Singapore', 'Jaipur', 'India/US',
       'Nagpur', 'Indore', 'New York, Bengaluru', 'California', 'India',
       'Ahemadabad', 'Rourkela', 'Bhubneswar', 'Chandigarh',
       'Delhi & Cambridge', 'Kolkata', 'Coimbatore', 'Bangalore',
       'Udaipur', 'Ahmedabad', 'Surat', 'Goa', 'Gaya', 'Vadodara',
       'Missourie', 'Panaji', 'Gwalior', 'Bangalore/ Bangkok', 'Kerala',
       nan, 'Lucknow', 'Trivandrum', 'SFO / Bangalore', 'Pune/Seattle',
       'Pune / Dubai', 'Bangalore / SFO', 'Varanasi', 'New Delhi / US',
       'Mumbai / UK', 'Kanpur

In [34]:
df['city'].isnull().sum()

4

In [35]:
df[df['city'].isnull()]

date      startup           industry  \
1600 2016-06-13  Matrubharti  Consumer Internet   
1908 2016-02-02   MaaxMarket  Consumer Internet   
2632 2015-07-23       Mswipe                NaN   
2736 2015-06-24        Indix                NaN   

                                subvertical city  \
1600                       e-Book Publisher  NaN   
1908          Marketing Automation platform  NaN   
2632  Mobile Point of Sale payment solution  NaN   
2736     SaaS product intelligence platform  NaN   

                                              investors investment_type  \
1600                                   Viridian Capital    Seed Funding   
1908  Multiple investors through Ten Minute Million ...    Seed Funding   
2632  Falcon Edge Capital, Meru Capital, Ola, Matrix...  Private Equity   
2736  Nokia Growth Partners, Nexus Venture Partners,...  Private Equity   

        amount  
1600     30000  
1908     22500  
2632  25000000  
2736  15000000

In [36]:
# Filling up the nan values with their corresponding cities
# Use this method for assignment it will not raise 'SettingWithCopyWarning'
# This happens because you are using chained indexing (df['col'][condition]), which can sometimes operate on a temporary copy of the data.
df.loc[df['city'].isnull(),'city']=['Ahemdabad','Chennai','Mumbai','Chennai']

In [37]:
# Only keeping the first city in case of entries with city1/city2
df['city'] = df['city'].str.split('/').apply(lambda x: x[0].strip())

In [38]:
df['city'].unique()

array(['Bengaluru', 'Gurgaon', 'New Delhi', 'Mumbai', 'Chennai', 'Pune',
       'Noida', 'Faridabad', 'San Francisco', 'San Jose,', 'Amritsar',
       'Delhi', 'Kormangala', 'Tulangan', 'Hyderabad', 'Burnsville',
       'Menlo Park', 'Gurugram', 'Palo Alto', 'Santa Monica', 'Singapore',
       'Taramani', 'Andheri', 'Chembur', 'Nairobi', 'Haryana', 'New York',
       'Karnataka', 'Bhopal', 'Bengaluru and Gurugram', 'India', 'Jaipur',
       'Nagpur', 'Indore', 'New York, Bengaluru', 'California',
       'Ahemadabad', 'Rourkela', 'Bhubneswar', 'Chandigarh',
       'Delhi & Cambridge', 'Kolkata', 'Coimbatore', 'Bangalore',
       'Udaipur', 'Ahmedabad', 'Surat', 'Goa', 'Gaya', 'Vadodara',
       'Missourie', 'Panaji', 'Gwalior', 'Kerala', 'Ahemdabad', 'Lucknow',
       'Trivandrum', 'SFO', 'Varanasi', 'Kanpur', 'Jodhpur', 'Boston',
       'USA', 'Belgaum', 'US', '\\\\xc2\\\\xa0Noida',
       '\\\\xc2\\\\xa0Bangalore', '\\\\xc2\\\\xa0New Delhi',
       '\\\\xc2\\\\xa0Gurgaon', '\\\\xc2\\\

In [39]:
# replacing the encoding \\\\xc2\\\\xa0 
df['city'] = df['city'].str.replace('\\\\xc2\\\\xa0','')

In [40]:
# correcting city names
np.sort(df['city'].unique())

array(['Ahemadabad', 'Ahemdabad', 'Ahmedabad', 'Amritsar', 'Andheri',
       'Bangalore', 'Belgaum', 'Bengaluru', 'Bengaluru and Gurugram',
       'Berlin', 'Bhopal', 'Bhubneswar', 'Boston', 'Burnsville',
       'California', 'Cape Town', 'Chandigarh', 'Chembur', 'Chennai',
       'Coimbatore', 'Dallas', 'Delhi', 'Delhi & Cambridge', 'Faridabad',
       'Gaya', 'Goa', 'Gurgaon', 'Gurugram', 'Gwalior', 'Haryana',
       'Hyderabad', 'India', 'Indore', 'Jaipur', 'Jodhpur', 'Kanpur',
       'Karnataka', 'Kathmandu', 'Kerala', 'Kolkata', 'Kormangala',
       'London', 'Los Angeles', 'Lucknow', 'Madrid', 'Menlo Park',
       'Missourie', 'Mumbai', 'Nagpur', 'Nairobi', 'New Delhi',
       'New York', 'New York, Bengaluru', 'Noida', 'Palo Alto', 'Panaji',
       'Pune', 'Rourkela', 'SFO', 'San Diego', 'San Francisco',
       'San Jose', 'San Jose,', 'Santa Monica', 'Silchar', 'Singapore',
       'Surat', 'Taramani', 'Tokyo', 'Toronto', 'Trivandrum', 'Tulangan',
       'US', 'USA', 'Udaipur', 

In [41]:
df['city'] = df['city'].replace(['Ahemadabad','Ahemdabad','Andheri','Bangalore','Bengaluru and Gurugram'],['Ahmedabad','Ahmedabad','Mumbai','Bengaluru','Bengaluru'])

In [42]:
df['city'] = df['city'].replace(['Delhi','Delhi & Cambridge','New York, Bengaluru'],['New Delhi','New Delhi','New York'])

In [43]:
df['city'] = df['city'].replace(['Gurgaon','Kormangala','SFO','San Jose,','US'],['Gurugram','Bengaluru','San Francisco','San Jose','USA'])

<h3>investment_type column</h3>

In [44]:
print('Unique values:\n',df['investment_type'].unique(),'\n')
print('Number of Null values',df['investment_type'].isnull().sum())

Unique values:
 ['Private Equity Round' 'Series C' 'Series B' 'Pre-series A' 'Seed Round'
 'Series A' 'Series D' 'Seed' 'Series F' 'Series E' 'Debt Funding'
 'Series G' 'Series H' 'Seed Funding' nan 'Funding Round' 'Maiden Round'
 'pre-series A' 'Seed Funding Round' 'Single Venture' 'Corporate Round'
 'Venture Round' 'Pre-Series A' 'Angel' 'Series J' 'Angel Round'
 'pre-Series A' 'Venture - Series Unknown' 'Private Equity'
 'Debt and Preference capital' 'Inhouse Funding' 'Seed/ Angel Funding'
 'Debt' 'Equity' 'Debt-Funding' 'Mezzanine' 'Series B (Extension)'
 'Equity Based Funding' 'Private Funding' 'Seed / Angel Funding'
 'Seed/Angel Funding' 'Seed / Angle Funding' 'Private'
 'Angel / Seed Funding' 'Structured Debt' 'Term Loan' 'PrivateEquity'
 'Seed\\\\nFunding' 'Private\\\\nEquity' 'Crowd funding' 'Crowd Funding'] 

Number of Null values 3


In [45]:
df[df['investment_type'].isnull()]

date                   startup     industry  \
30 2019-11-19  Furtados School of Music    Education   
76 2019-06-03                 FabHotels   E-Commerce   
81 2019-06-06               Sistema.bio  Agriculture   

                   subvertical      city  \
30             Music Education  Tulangan   
76                 Hospitality  Gurugram   
81  Hybrid Reactor Biodigestor   Nairobi   

                                            investors investment_type  \
30                 IAN Fund and DSG Consumer Partners             NaN   
76         Goldman Sachs, Accel Partners and Qualcomm             NaN   
81  Shell Foundation, DILA CAPITAL, Engie RDE Fund...             NaN   

       amount  
30  200000000  
76    4889976  
81    2739035

In [46]:
# Filling up NaN values with 'unknown'
df['investment_type'] = df['investment_type'].fillna('unknown')

In [47]:
# dealing with encoding issues
df['investment_type'] = df['investment_type'].str.replace('\\\\n',' ')

In [48]:
# Correcting the names

In [49]:
df['investment_type'] = df['investment_type'].str.title()
df['investment_type'] = df['investment_type'].replace(['Angel'], ['Angel Round'])

In [50]:
df['investment_type'] = df['investment_type'].replace(['Angel / Seed Funding','Seed / Angel Funding','Seed / Angle Funding','Seed/ Angel Funding'],'Seed/Angel Funding')

In [51]:
df['investment_type'] = df['investment_type'].replace(['Debt', 'Debt-Funding'],'Debt Funding')

In [52]:
df['investment_type'] = df['investment_type'].replace('Equity Based Funding','Equity')

In [53]:
df['investment_type'] = df['investment_type'].replace(['Private Equity Round'],'Private Equity')

In [54]:
df['investment_type'] = df['investment_type'].replace(['Seed', 'Seed Funding Round', 'Seed Round'],'Seed Funding')

In [55]:
df['investment_type'] = df['investment_type'].replace('Venture - Series Unknown','Venture Round')

In [56]:
print('Unique values:\n',np.sort(df['investment_type'].unique()),'\n')

Unique values:
 ['Angel Round' 'Corporate Round' 'Crowd Funding'
 'Debt And Preference Capital' 'Debt Funding' 'Equity' 'Funding Round'
 'Inhouse Funding' 'Maiden Round' 'Mezzanine' 'Pre-Series A' 'Private'
 'Private Equity' 'Private Funding' 'Privateequity' 'Seed Funding'
 'Seed/Angel Funding' 'Series A' 'Series B' 'Series B (Extension)'
 'Series C' 'Series D' 'Series E' 'Series F' 'Series G' 'Series H'
 'Series J' 'Single Venture' 'Structured Debt' 'Term Loan' 'Unknown'
 'Venture Round'] 



<H3>startup column</H3>

In [57]:
df['startup'].nunique()

1703

In [58]:
df['startup'] = df['startup'].str.lower().str.strip()

In [59]:
# removing .com, .in, pvt ltd from names
df['startup'] = df['startup'].str.replace('.com','')
df['startup'] = df['startup'].str.replace('.in','')
df['startup'] = df['startup'].str.replace('pvt ltd.','')
df['startup'] = df['startup'].str.replace('pvt. ltd.','')
df['startup'] = df['startup'].str.replace('pvt. ltd','')
df['startup'] = df['startup'].str.replace('ltd.','')
df['startup'] = df['startup'].str.replace('inc.','')
df['startup'] = df['startup'].str.replace('inc','')
df['startup'] = df['startup'].str.replace('https://www.wealthbucket/','wealthbucket')

In [60]:
df['startup'].nunique()

1615

In [61]:
# removing the encodings
# \xc2\xa0News in shorts, 
# NearBuy (previously\xc2\xa0 groupon India), 
# Carl\xe2\x80\x99s Jr, 
# Byju\xe2\x80\x99s 
# \xc2\xa0CloudCherry 
# Let\xe2\x80\x99s Barter 
# "BYJU\'S" 
# Alef\nMobitech

In [62]:
df['startup'] = df['startup'].str.replace(r'\\xe2\\x80\\x99',"'")
df['startup'] = df['startup'].str.replace(r'\\xc2\\xa0',"")

In [63]:
df['startup'] = df['startup'].str.replace('\xc2\xa0',"")
df['startup'] = df['startup'].str.replace('\xe2\x80\x99',"'")
df['startup'] = df['startup'].str.replace('\\',"")
df['startup'] = df['startup'].str.replace('\\n'," ")
df['startup'] = df['startup'].str.replace('"','')

In [64]:
# df['startup'][df['startup'].str.contains('oyo')]

In [65]:
df['startup'].nunique()

1613

In [66]:
def select_name(name):
    if name=='the/nudge foundation':
        return name
    else:
        return name.split('/')[0].strip()
    

In [67]:
# selecting the new name of the startup from entries like transtutor/askiitians
df['startup'] = df['startup'].apply(select_name)

In [68]:
# Combaining names based on their phenotics
!pip install jellyfish

# Many startup names have different spellings  

In [69]:
import jellyfish
from thefuzz import fuzz
from collections import defaultdict

# Create a dictionary to store names and their phonetic codes
# We'll use a defaultdict to easily group names by their code
phonetic_groups=defaultdict(list)

# Get unique startup names
unique_names=df['startup'].unique()

# Phonetic Encoding
for name in unique_names:
    # generating the metaphone code for each name
    metaphone = jellyfish.metaphone(name)
    # so that names with spaces and non-spaced will have the same code like oyorooms and oyo rooms
    metaphone = metaphone.replace(' ','')
    # grouping names based on the metaphone code
    phonetic_groups[metaphone].append(name)

# fuzzy string matching
similar_name_groups=[]
for group, names in phonetic_groups.items():
    # We will only check for groups with 2 or more names
    if len(names)>1:
        # Keep track of names that have already been grouped
        matched_names=set()
        # to collect all the similar names in a group
        temp_group=[]
        
        for i in range(len(names)):
            if names[i] in matched_names:
                continue
            # to collect similar names in one iteration
            current_group=[names[i]]
            for j in range(i+1,len(names)):
                if names[j] in matched_names:
                    continue
                fuzz_score = fuzz.token_set_ratio(names[i],names[j])
                if fuzz_score>80:
                    matched_names.add(names[j])
                    current_group.append(names[j])
            if len(current_group)>1:
                temp_group.extend(current_group)
        if len(temp_group)>1:
            similar_name_groups.append(temp_group)
        
for group in similar_name_groups:
    print(group)

['byju’s', "byju's"]
['aye finance', 'ayefinance']
['fabhotels', 'fab hotels']
['avail finance', 'availfinance']
['unacademy', 'unaacademy']
['oyorooms', 'oyo rooms']
['guiddoo', 'guidoo']
['letstransport', 'lets transport']
['daily hunt', 'dailyhunt']
['loan tap', 'loantap']
['zippserv', 'zippserve']
['mengage', 'moengage']
['book my show', 'bookmyshow']
['entropik', 'entropika']
['cure fit', 'curefit']
['traveltriangle', 'travel triangle']
['wellthy', 'wealthy']
['earlysalary', 'early salary']
['logic roots', 'logicroots']
['tempgo', 'tempogo']
['bank bazaar', 'bankbazaar']
['credit vidya', 'creditvidya']
['transerve', 'transerv']
['ink monk', 'inkmonk']
['rentomojo', 'rentmojo']
['bhive workspace', 'bhiveworkspace']
['legal raasta', 'legalraasta']
['gennext students', 'genextstudents']
['faasos', "faaso's"]
['intelligence node', 'intelligencenode']
['india lends', 'indialends']
['shadowfox', 'shadowfax']
['ue lifesciences', 'ue life sciences']
['icustommadeit', 'icustomadeit']


In [70]:
for group in similar_name_groups:
    group.sort()
    df['startup']=df['startup'].replace(group[1],group[0])

<h3>investors column</h3>

In [71]:
df['investors'].nunique()

1666

In [72]:
# For simplicity keeping only the first investor name 
import math
def first_investor(name):
    # for single values we use math.isnan() to check for nan values
    if isinstance(name, str):
        return name.split(',')[0].strip()
    else:
        return name

df['investors'] = df['investors'].apply(first_investor)

In [73]:
# Total number of nan values
df['investors'].isnull().sum()

20

In [74]:
# removing the nan values 
df.dropna(subset=['investors'],inplace=True)

In [75]:
df['investors'].isnull().sum()

0

In [76]:
df['investors'].nunique()

1245

In [77]:
# removing the encodings
df['investors'] = df['investors'].str.replace(r'\\xe2\\x80\\x99',"'")
df['investors'] = df['investors'].str.replace(r'\\xc2\\xa0',"")
df['investors'] = df['investors'].str.replace('\\',"")
df['investors'] = df['investors'].str.replace('\\n'," ")

In [78]:
df['investors'][df['investors'].str.contains(r'\\')]

Series([], Name: investors, dtype: object)

<H3>industry column</H3>

In [79]:
# Total nan values in the industry column
df['industry'].isnull().sum()

649

In [80]:
# removing the encodings
df['industry'] = df['industry'].str.replace(r'\\xe2\\x80\\x99',"'")
df['industry'] = df['industry'].str.replace(r'\\xc2\\xa0',"")
df['industry'] = df['industry'].str.replace('\\',"")
df['industry'] = df['industry'].str.replace('\\n'," ")

In [81]:
df['industry'].nunique()

94

In [82]:
# changing the industry column to lower case
df['industry'] = df['industry'].str.lower()

In [83]:
df['industry'].nunique() # unique values reduced

83

<h3>subvertical column</h3>

In [84]:
# Total nan values in the industry column
df['subvertical'].isnull().sum()

131

In [85]:
# removing the encodings
df['subvertical'] = df['subvertical'].str.replace(r'\\xe2\\x80\\x99',"'")
df['subvertical'] = df['subvertical'].str.replace(r'\\xc2\\xa0',"")
df['subvertical'] = df['subvertical'].str.replace('\\',"")
df['subvertical'] = df['subvertical'].str.replace('\\n'," ")

In [86]:
df['subvertical'].nunique()

1759

<h3>filling up nan values in industry and subvertical using genai</h3>

In [87]:
# !pip install -q -U google-genai

In [88]:
# importing the required libraries
from google import genai
import time
import os
client = genai.Client(api_key="AIzaSyA81-lxs5T8UfEC_6cny2jdV7CeIw_WvbE")

# --- Function to get info from the API ---
def get_startup_info(startup_name):
    # prompt
    prompt = f"""
    Based on your knowledge, what is the primary Industry and Sub-Vertical for the startup named "{startup_name}"?
    Please provide the answer in the format: Industry; Sub-Vertical.
    For example, for 'Stripe', the answer should be 'FinTech; Payment Gateway'.
    If you cannot determine the industry, please return 'Unknown; Unknown'.
    """
    try:
        response = client.models.generate_content(model="gemini-2.5-flash", contents=prompt)
        # Clean up the response text
        parts = response.text.strip().split(';')
        if len(parts) == 2:
            return parts[0].strip(), parts[1].strip()
        else:
            return "Failed to Parse", "Failed to Parse"
    except Exception as e:
        print(f"An error occurred for {startup_name}: {e}")
        return "API Error", "API Error"

In [104]:
# creating a series of 25 startup names where industry is null
startup_null_industry= df['startup'][df['industry'].isnull()].head(25)
num_nan_values = df['industry'].isnull().sum()
print(f"Remaining NaN values in industry column: {num_nan_values}")
results = []
i=0
for index, row in startup_null_industry.items():
    name = row
    print(f"Processing ({i + 1}/{len(startup_null_industry)}): {name}...")
        
    industry, sub_vertical = get_startup_info(name)
        
    if industry=='API Error':
        i=i+1
        continue
        
    if industry=='Failed to Parse':
        results.append({
            'Startup Name': name,
            'Industry': 'unknown',
            'Sub-Vertical': 'unknown'})
        df.loc[index,'industry']='unknown'
        df.loc[index, 'subvertical']='unknown'
    else:
        results.append({
                'Startup Name': name,
                'Industry': industry,
                'Sub-Vertical': sub_vertical
            })
        df.loc[index,'industry']=industry
        df.loc[index, 'subvertical']=sub_vertical
        
    time.sleep(1) # Pause for 1 second between requests
    i=i+1
    
num_nan_values = df['industry'].isnull().sum()
print(f"Remaining NaN values in industry column: {num_nan_values}")

Remaining NaN values in industry column: 341
Processing (1/25): reverie...
Processing (2/25): cloudcherry analytics...
Processing (3/25): wudstay...
Processing (4/25): flytxt...
Processing (5/25): axtria...
Processing (6/25): 33coupons...
Processing (7/25): hopping chef...
Processing (8/25): collectabillia...
Processing (9/25): lendingkart...
Processing (10/25): lybrate...
Processing (11/25): zoomcar...
Processing (12/25): news in shorts...
Processing (13/25): bluestone...
An error occurred for bluestone: 429 RESOURCE_EXHAUSTED. {'error': {'code': 429, 'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits.', 'status': 'RESOURCE_EXHAUSTED', 'details': [{'@type': 'type.googleapis.com/google.rpc.QuotaFailure', 'violations': [{'quotaMetric': 'generativelanguage.googleapis.com/generate_content_free_tier_requests', 'quotaId': 'GenerateRequestsPerDayPerProjectPerMo

In [90]:
# for row in results:
#     print(f"Processing: {row['Startup Name']}...")
#     df.loc[df['startup']==row['Startup Name'],'industry']=row['Industry']
#     df.loc[df['startup']==row['Startup Name'],'industry']=row['Industry']
#     print(f'Processing complete for: {row['Startup Name']}')

In [105]:
df.to_csv('cleaned_startup_funding.csv')